kube/src/client/mod.rs

```diff
@@ -77,97 +77,79 @@ impl Client {
         Self::try_from(client_config)
     }
 
-    async fn send(&self, request: http::Request<Vec<u8>>) -> Result<reqwest::Response> {
-        let (parts, body) = request.into_parts();
+    async fn send(&self, request: Request<Body>) -> Result<Response<Body>> {
+        let (mut parts, body) = request.into_parts();
         let pandq = parts.uri.path_and_query().expect("valid path+query from kube");
         let uri_str = finalize_url(&self.cluster_url, &pandq);
+        parts.uri = uri_str.parse().expect("valid URL");
         //trace!("Sending request => method = {} uri = {}", parts.method, &uri_str);
-
-        let mut headers = parts.headers;
+        let mut headers = self.headers.clone();
+        headers.extend(parts.headers.clone().into_iter());
         // If we have auth headers set, make sure they are updated and attached to the request
-        if let Some(auth_header) = self.config.get_auth_header().await? {
-            headers.insert(reqwest::header::AUTHORIZATION, auth_header);
+        if let Some(auth_header) = self.auth_header.to_header().await? {
+            headers.insert(http::header::AUTHORIZATION, auth_header);
         }
+        parts.headers = headers;
 
         let request = match parts.method {
             http::Method::GET
             | http::Method::POST
             | http::Method::DELETE
             | http::Method::PUT
-            | http::Method::PATCH => self.inner.request(parts.method, &uri_str),
+            | http::Method::PATCH => Request::from_parts(parts, body),
             other => return Err(Error::InvalidMethod(other.to_string())),
         };
 
-        let req = request.headers(headers).body(body).build()?;
-        let res = self.inner.execute(req).await?;
+        let res = self.inner.request(request).await?;
```

Comment:
Every request goes through this `self.inner.request` now.

We should be able to change `kube::Client` to use `tower::Service` instead of `hyper::Client` (`self.inner.request` becomes `self.inner.call`). Then we should be able to extract all the changes to `Request` in this method out of `kube::Client` to composable layers created from `kube::Config`.

- Set `cluster_url`
- Add default `headers`
- Set auth header
- Validate method
- Support compression
- Layer for log/trace

Reply:
Wow. This is fantastic.

Where do you envision the underlying `hyper::Client` will end up after this? Inside `kube::Config` (since it might need to do oauth http calls)? Inside user code, and passed to `Config`?

Reply:
I'm thinking of making `kube::Client` take `tower::Service`.  We'll provide `kube::Service` (based on `hyper::Client`) and a convenient method to be used by default to construct a client. Users can use their implementation or decorate `kube::Service` if they need something not supported out of the box. This will make it easier to test as well.

The `hyper::Client` used in `oauth2` is a separate instance because it can't use the custom `HttpsConnector`.

---

`hyper::Client` is `tower::Service`, so `kube::Client` should just work after changing types, but I didn't realize how much change is required by adding a type parameter to `kube::Config`. I need to study `kube::api` more.

Making this fit naturally should be the last thing to experiment with. If it works well enough, I'll open a new PR for merging.
We shouldn't lose anything and the WebSocket connection upgrade is huge, so this is looking so much better than I was originally thinking.

Reply:
Yeah, the one client and one send method is really making things clearer. Plan makes sense to me.